### ISSUE TITLE
Replace trait object returns and parameters with optionally unsized generics

### ISSUE THREAD

**Issue by ok-nick** [veeso/remotefs-rs#7]
The `RemoteFs` trait accepts boxed trait objects and returns boxed trait objects for many of its methods. This can be slightly inconvenient when working with many of its methods, forcing you to box your struct before passing it through. Additionally, it also adds a slight overhead with an unnecessary vtable in many cases.

Some protocols may force you to deal with unsized types (is this actually true?), so it could be optionally unsized. Something along the lines of,
```rs
fn on_read<T: Read + ?Sized>(&mut self, _readable: T) -> RemoteResult<()>
```

**Comment by veeso** [comment]
The box thing has been implemented like that on purpose, since otherwise you wouldn't be allowed to use trait objects, like:

```rust
let client: Box<dyn RemoteFs> = Box::new(MockRemoteFs {});
```

**Comment by veeso** [comment]
May be fixed by issue #9
